### PR TITLE
add ShoppingCartController, AddToCartRequestDto, CartItemDto, CartIte…

### PR DESCRIPTION
**Pull request by Alexander1iubchenko**
add ShoppingCartController, AddToCartRequestDto, CartItemDto, CartItemRequestUpdateDto, ShoppingCartDto, CartItemMapper, ShoppingCartMapper, CartItem, ShoppingCart, CartItemRepository, ShoppingCartRepository, ShoppingCartService, ShoppingCartServiceImpl,11-create-shoppingcarts-table.yaml, 12-create-cartitems-table.yaml

### CODE REVIEW

File: src/main/java/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import service.ShoppingCartService;
+
+@RestController
+@RequestMapping("/api/cart")
```
Comment (BohdanTrue) [inline src/main/java/controller/ShoppingCartController.java:24]: It's not necessary but you can add this property `server.servlet.context-path=/api` to application.properties and write all RequestMapping's without prefix `/api`

File: src/main/java/model/Book.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ public class Book {
     private String description;
     @Column(name = "cover_image")
     private String coverImage;
-    @Column(nullable = false)
+    @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
```
Comment (BohdanTrue) [inline src/main/java/model/Book.java:54]: In my project Hibernate create this column by default with right naming, without explicit indication

image

File: src/main/java/service/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addToCart(AddToCartRequestDto requestDto, Long userId) {
+        Book book = bookRepository.findById(requestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException("Can not find book by this id:"
+                        + requestDto.getBookId()));
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can not find user by this id:"
```
Comment (BohdanTrue) [inline src/main/java/service/ShoppingCartServiceImpl.java:36]: make constants for messages

File: src/main/resources/db/changelog/changes/11-create-shoppingcarts-table.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        - createTable:
+            tableName: shopping_carts
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
Comment (BohdanTrue) [inline src/main/resources/db/changelog/changes/11-create-shoppingcarts-table.yaml:15]: unnecessary make nullable constraint for primaryKey
primaryKey has this (nullable) constraint by default

**Review by BohdanTrue — APPROVED** [review]
GJ

File: src/main/java/mapper/BookMapper.java [hunk 5]
```diff
@@ -21,6 +22,11 @@ public interface BookMapper {
     @Mapping(target = "categories", ignore = true)
     Book toModel(CreateBookRequestDto requestDto);
 
+    @Named("bookFromId")
+    default Book bookFromId(Long id) {
+        return null;
```
Comment (vovasalyha) [inline src/main/java/mapper/BookMapper.java]: Why does it always return null?

File: src/main/java/service/ShoppingCartService.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import dto.cart.AddToCartRequestDto;
+import dto.cart.CartItemRequestUpdateDto;
+import dto.cart.ShoppingCartDto;
+
+public interface ShoppingCartService {
+    ShoppingCartDto addToCart(AddToCartRequestDto requestDto, Long userId);
+
+    ShoppingCartDto getShoppingCartDto(Long userId);
+
+    ShoppingCartDto updateQuantityOfBooksInShoppingCart(Long userId, Long id,
```
Comment (vovasalyha) [inline src/main/java/service/ShoppingCartService.java]:
```suggestion
    ShoppingCartDto updateCartItem(Long userId, Long cartItemId,
```

File: src/main/java/service/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public ShoppingCartDto getShoppingCartDto(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can not find Shopping "
+                        + "cart with this user id" + userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/service/ShoppingCartServiceImpl.java:65]:
```suggestion
    @Override
    @Transactional
```

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments.

File: src/main/java/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    private Long id;
+    @Column(nullable = false)
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id")
+    private User user;
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @OneToMany(fetch = FetchType.EAGER, mappedBy = "shoppingCart")
```
Comment (d1sam) [inline src/main/java/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart")
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some minor changes still should be done :)

**Review by d1sam — APPROVED** [review]
well done!

**Pull request merged**